python/cudf_polars/cudf_polars/dsl/expr.py

```diff
@@ -0,0 +1,1023 @@
+        table = do_sort(plc.Table([column.obj]), order, null_order)
+        return Column(table.columns()[0], column.name).set_sorted(
+            is_sorted=plc.types.Sorted.YES, order=order[0], null_order=null_order[0]
+        )
+
+    def collect_agg(self, *, depth: int) -> AggInfo:
+        """Collect information about aggregations in groupbys."""
+        # TODO: Could do with sort-based groupby and segmented sort post-hoc
+        raise NotImplementedError("Sort in groupby")
+
+
+class SortBy(Expr):
+    __slots__ = ("options", "children")
+    _non_child = ("dtype", "options")
+
+    def __init__(
+        self,
+        dtype: plc.DataType,
+        options: tuple[bool, bool, tuple[bool]],
+        column: Expr,
+        *by: Expr,
+    ):
+        super().__init__(dtype)
+        self.options = options
+        self.children = (column, *by)
+
+    @with_mapping
+    def evaluate(
+        self,
+        df: DataFrame,
+        *,
+        context: ExecutionContext = ExecutionContext.FRAME,
+        mapping: dict[Expr, Column] | None = None,
+    ) -> Column:
+        """Evaluate this expression given a dataframe for context."""
+        column, *by = (
+            child.evaluate(df, context=context, mapping=mapping)
+            for child in self.children
+        )
+        (stable, nulls_last, descending) = self.options
+        order, null_order = sorting.sort_order(
+            descending, nulls_last=nulls_last, num_keys=len(by)
+        )
+        do_sort = plc.sorting.stable_sort_by_key if stable else plc.sorting.sort_by_key
+        table = do_sort(
+            plc.Table([column.obj]), plc.Table([c.obj for c in by]), order, null_order
+        )
+        return Column(table.columns()[0], column.name)
+
+    def collect_agg(self, *, depth: int) -> AggInfo:
+        """Collect information about aggregations in groupbys."""
+        # TODO: Could do with sort-based groupby and segmented sort post-hoc
+        raise NotImplementedError("SortBy in groupby")
+
+
+class Gather(Expr):
+    __slots__ = ("children",)
+    _non_child = ("dtype",)
+
+    def __init__(self, dtype: plc.DataType, values: Expr, indices: Expr):
+        super().__init__(dtype)
+        self.children = (values, indices)
+
+    @with_mapping
+    def evaluate(
+        self,
+        df: DataFrame,
+        *,
+        context: ExecutionContext = ExecutionContext.FRAME,
+        mapping: dict[Expr, Column] | None = None,
+    ) -> Column:
+        """Evaluate this expression given a dataframe for context."""
+        values, indices = (
+            child.evaluate(df, context=context, mapping=mapping)
+            for child in self.children
+        )
+        lo, hi = plc.reduce.minmax(indices.obj)
+        lo = plc.interop.to_arrow(lo).as_py()
+        hi = plc.interop.to_arrow(hi).as_py()
+        n = df.num_rows
+        if hi >= n or lo < -n:
+            raise ValueError("gather indices are out of bounds")
```

Comment:
I wonder if we should add a new `out_of_bounds_policy` `ERROR` or so to libcudf to support this use case. WDYT? It seems like something that probably has multiple use cases.

Reply:
I think that might be useful, let me think about how that is doable. To avoid a pre-pass over the gather indices one would need to flag the error in the `__device__` code that produces the indices. Since we can't raise an exception `__device__` kernels, we would need to have a flag that is post-hoc checked, which might serialise too much if atomic.

Reply:
Yeah, classically you'd do the UB thing where every thread sets a boolean without an atomic and trust that it'll always produce the right output, but we don't want that these days so you'd presumably need to do a reduction on the flag instead.